pkg/ruler/manager.go

```diff
@@ -100,7 +111,62 @@ func NewDefaultMultiTenantManager(cfg Config, managerFactory ManagerFactory, reg
 		}, []string{"user"}),
 		registry: reg,
 		logger:   logger,
-	}, nil
+	}
+
+	var err error
+	switch cfg.AlertManagerDiscovery.Mode {
+	case alertmanagerdiscovery.ModeRing:
+		level.Info(logger).Log("msg", "using ring based alert manager discovery")
+		m.discoveryService, err = alertmanager.NewRing(cfg.AlertManagerRing, "ruler", m, logger, reg)
+		if err != nil {
+			return nil, err
+		}
+
+	default:
+		level.Info(logger).Log("msg", "using dns based alert manager discovery")
+		dnsProvider := alertmanagerdiscovery.NewDNSProvider(reg)
+		err = alertmanagerdiscovery.BuildDiscoveryConfigs(cfg.AlertmanagerURL, m.discoveryConfigs, cfg.AlertmanagerRefreshInterval, dnsProvider)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	m.notifierCfg, err = buildNotifierConfig(&cfg, m.discoveryConfigs)
```

Comment:
From PR description:

> While this PR is similar to https://github.com/grafana/mimir/pull/2957, it doesn't create a sub-service for DNS discovery like the query scheduler does, since the existing ruler code used dns.NewProvider whereas the scheduler's discovery uses grpcutil.NewDNSResolverWithFreq.

Design wise, I'm not a big fan of doing the same thing in two different ways in this PR: the ring-based discovery calls InstanceAdded/Removed() and then builds the notifier config, while the DNS-based discovery has a static notifier config which implements `NewDiscoverer()`.

Question: can't we just have the ring-based ring implementing `NewDiscoverer()` too?

Reply:
I think I see what you mean. While `NewDiscoverer` is created the same way in either DNS or ring mode, via `rulerNotifier.applyConfig` (indirectly), the `discovery.Configs` that method uses are created in different ways and at different times.

One thing we could do is combine the `alertmanagerdiscovery.BuildDiscoveryConfigs` (renamed to `NewDiscoveryConfigs`) and `alertmanagerdiscovery.NewDiscoveryConfig` functions, where right now the latter is intended just for DNS based configs. The combined function would still be called at different times though - for DNS it would be called when we first instantiate the `DefaultMultiTenantManager`, and for ring it would be called as instances are added/removed. I'm not sure if this is what you meant though.